include/lbann/io/persist.hpp

```diff
@@ -94,27 +98,80 @@ class persist {
-  bool read_int32_contig (persist_type type, const char *name, int32_t *buf, uint64_t count);
 
-  bool write_float(persist_type type, const char *name, float  val);
-  bool read_float (persist_type type, const char *name, float *val);
+  template<typename T>
+  bool write_parameter_vector(persist_type type, const char *name, std::vector<T> val, int array_size) {
+    return write_bytes(type, name, val.data(), array_size * sizeof(T));
+  }
+  
+  template<typename T>
+  bool read_parameter_vector(persist_type type, const char *name, std::vector<T> &val, int array_size) {
+    return read_bytes(type, name, val.data(), array_size * sizeof(T));
+  }
 
   bool write_string(persist_type type, const char *name, const char *val, int str_length);
   bool read_string (persist_type type, const char *name, char *val, int str_length);
 
-  bool write_double(persist_type type, const char *name, double  val);
-  bool read_double (persist_type type, const char *name, double *val);
+  
+  #ifdef LBANN_HAS_HDF5   
+  
+  H5::PredType cpp_to_hdf5(int val) { return H5::PredType::NATIVE_INT; } 
+  H5::PredType cpp_to_hdf5(double val) { return H5::PredType::NATIVE_DOUBLE; }
+  H5::PredType cpp_to_hdf5(float val) { return H5::PredType::NATIVE_FLOAT; }
+  H5::PredType cpp_to_hdf5(long val) { return H5::PredType::NATIVE_LONG; }
+  
+  template<typename T>
+  bool write_hdf5_parameter(H5::Group group_name, const char *name, T *val) {
+    H5::DataSpace dataspace = H5::DataSpace();
+    H5::PredType hdf5_type = cpp_to_hdf5(*val);
+    H5::Attribute attribute = group_name.createAttribute(name, hdf5_type, dataspace);
+    attribute.write(hdf5_type, static_cast<void*>(val));
+    return true;
+  }
+
+
+  template<typename T>
+  bool read_hdf5_parameter(H5::Group group_name, const char *name, T *val) {
+    H5::Attribute attr = group_name.openAttribute(name);
+    H5::DataType type = attr.getDataType();
+    attr.read(type,val);     
+    return true;
+  }
+
+  template<typename T>
+  bool write_hdf5_array(H5::Group group_name, const char *name, std::vector<T> val) {
+    const hsize_t arr_size = val.size();
+    H5::PredType hdf5_type = cpp_to_hdf5(val[0]);
+    H5::DataSpace dataspace = H5::DataSpace(1,&arr_size);
+    H5::DataSet dataset = group_name.createDataSet(name, hdf5_type, dataspace);
+    dataset.write(val.data(), hdf5_type);
+    return true;
+  }
+ 
+ 
+  template<typename T>
+  bool read_hdf5_array(H5::Group group_name, const char *name, std::vector<T> &val) {
+    H5::DataSet ds = group_name.openDataSet(name);
+    H5::DataSpace dataspace= ds.getSpace();
+    ds.read(val.data(), H5::PredType::NATIVE_INT, dataspace);
+    return true;
+  }
 
-  bool write_datatype(persist_type type, const char *name, DataType  val);
-  bool read_datatype (persist_type type, const char *name, DataType *val);
+  H5::Group getGroup(std::string group_name);
+  bool write_hdf5_distmat(std::string group_name, const char *name, AbsDistMat *M, lbann_comm *comm);
```

Comment:
You can make comm pointer "const lbann_comm*" or simply replace it with "bool master" and pass comm->am_world_master()